### PR TITLE
Plumbing and core MoE logic for router replay

### PR DESCRIPTION
**Pull request by xuefgu**
# Description

This PR introduces forced expert routing in MaxText’s MoE blocks to support RL training (e.g., GRPO). In this pipeline, completions are generated on vLLM, and their exact expert routing selections must be strictly enforced during training on MaxText to prevent expert selection mismatch and policy gradient divergence.

The externally supplied 4D routing tensor ([bs, seq_len, num_layers, k]) enters via TunixMaxTextAdapter and the top-level Transformer model. It is then sliced layer-by-layer inside decoders.py (Linen) or nnx_decoders.py (NNX) and routed down through model-specific decoder layers (DeepSeek, Qwen3, Qwen3.5, Gemma4, and Mixtral) into the core MoE block (moe.py), where it overrides the model-determined expert selections.

The implementation also implements out-of-bounds safety by mapping vLLM's padding sentinel values (-1) to valid, evenly distributed dummy expert indices before calling JAX bincount in permute(), ensuring hardware load balancing and performance. It also preserves backpropagation to router weights by executing manual gating logit extraction based on the forced selections.

This PR contains the core plumbing, edge case handling, and unscanned decoder loop support. A follow-up PR will support forced expert routing for scan_layers=True.

# Tests

Unit tests.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run end-to-end tests tests and provided workload links above if applicable.
- [X] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

**Comment by shralex** [comment]
@gemini-cli /investigate

File: src/maxtext/layers/decoders.py [hunk 1]
```diff
@@ -1121,19 +1131,38 @@ def __call__(
...
-                **layer_call_kwargs,
-            )
+            current_forced_routed_experts = None
+            is_moe = False
+            if cfg.decoder_block in (
+                DecoderBlockType.MIXTRAL,
+                DecoderBlockType.QWEN3_MOE,
+                DecoderBlockType.QWEN3_NEXT,
+                DecoderBlockType.QWEN3_5,
+                DecoderBlockType.QWEN3_CUSTOM_MOE,
```
Comment (gagika) [inline src/maxtext/layers/decoders.py:1141]: how about DeepSeek or Kimi that have both MoE layers and Dense?

Also Gemma4 has 2 models MoE and Dense, but handled by the same layer DecoderBlockType.Gemma4 

For Gemma4 you can have:  DecoderBlockType.GEMMA4 and num_experts > 1

File: src/maxtext/layers/moe.py [hunk 2]
```diff
@@ -595,32 +595,47 @@ def should_update_load_balance(self):
...
 
     if self.config.decoder_block == ctypes.DecoderBlockType.DEEPSEEK:
       top_k_weights = self.deepseek_scale_weights(top_k_weights)
     elif self.config.decoder_block not in (ctypes.DecoderBlockType.LLAMA4, ctypes.DecoderBlockType.GEMMA4):
       top_k_weights = jax.nn.softmax(top_k_weights.astype(jnp.float32), axis=-1).astype(self.dtype)
 
+    # Zero out weights for padding indices!
+    if forced_routed_experts is not None:
+      valid_token_mask = top_k_indices[:, :, 0] != -1
+      top_k_weights = top_k_weights * valid_token_mask[:, :, None]
```
Comment (gagika) [inline src/maxtext/layers/moe.py:637]: why not

```
valid_token_mask = top_k_indices != -1
top_k_weights = top_k_weights * valid_token_mask
```

like above?
